src/io/config_auto.cpp

```diff
@@ -673,6 +673,7 @@ std::string Config::SaveMembersToString() const {
   str_buf << "[max_depth: " << max_depth << "]\n";
   str_buf << "[min_data_in_leaf: " << min_data_in_leaf << "]\n";
   str_buf << "[min_sum_hessian_in_leaf: " << min_sum_hessian_in_leaf << "]\n";
+  str_buf << "[data_sample_strategy: " << data_sample_strategy << "]\n";
```

Comment:
This file is automatically generated and isn't intended to be edited manually.

https://github.com/microsoft/LightGBM/blob/e0d63b573505c5ee6c1ea8dd34154b4b46d5981b/src/io/config_auto.cpp#L6

If you want to propose adding `data_sample_strategy` in the model file, you should do that by removing this line:

https://github.com/microsoft/LightGBM/blob/e0d63b573505c5ee6c1ea8dd34154b4b46d5981b/include/LightGBM/config.h#L163

And then running the following from the root of the repo:

```python
python helpers/parameter_generator.py
```

But also... why is it important to have this in the model file? @shiyu1994 was this intentionally kept out of the model file in #4827 or was that just an oversight?

Reply:
If someone is able to regenerate this file and pick up the equivalent change, that would be great, as I don't use Python.

Reply:
I'm not sure why this conversation was marked resolved.

@mjmckp I'd still like you to answer my question:

> *why is it important to have this in the model file?*

and for @shiyu1994 to answer my other question

> *was this intentionally kept out of the model file in https://github.com/microsoft/LightGBM/pull/4827 or was that just an oversight?*

Reply:
I don't think sampling strategy should be in the model file. It is actually part of the training algorithm, instead of the model. Just like we don't include the boosting type in the model file. (Actually, before the refactor in #4827, sampling strategy `goss` is a boosting type).

Reply:
Thank you! I agree with that.

Reply:
@jameslamb @shiyu1994  The reason I added `data_sample_strategy` to the model file is because LightGBM.Net needs to  validate that the parameters found in the model file match those presented at the input to training.  Every other parameter that affects the training is present in the file (despite the value of doc_only in config.h btw), and without this change I get an exception that there is a parameter mismatch for this one parameter only.  What do you suggest we do?

Reply:
> Every other parameter that affects the training is present in the file (despite the value of doc_only in config.h btw)

Can you provide a *specific* example of a parameter you see in model files geberated by LightGBM which is marked as doc-only?

Reply:
I've just grabbed a random example from the unit tests for LightGBMNet (using LightGBM built with the changes in this PR).

The input parameter string is:
```
verbosity=0 linear_tree=true is_enable_sparse=false use_missing=false precise_float_parser=true min_data_in_leaf=3 monotone_constraints=0,-1,-1,-1,0,0,-1,0 objective=binary sigmoid=0.66 metric=binary_logloss metric_freq=3 boosting=dart data_sample_strategy=goss num_iterations=2 learning_rate=0.045
```

The parameters in the model file are:
```
parameters:
[boosting: dart]
[objective: binary]
[metric: binary_logloss]
[tree_learner: serial]
[device_type: cpu]
[data: ]
[valid: ]
[num_iterations: 2]
[learning_rate: 0.045]
[num_leaves: 31]
[num_threads: 0]
[deterministic: 0]
[force_col_wise: 0]
[force_row_wise: 0]
[histogram_pool_size: -1]
[max_depth: -1]
[min_data_in_leaf: 3]
[min_sum_hessian_in_leaf: 0.001]
[data_sample_strategy: goss]
[bagging_fraction: 1]
[pos_bagging_fraction: 1]
[neg_bagging_fraction: 1]
[bagging_freq: 0]
[bagging_seed: 3]
[feature_fraction: 1]
[feature_fraction_bynode: 1]
[feature_fraction_seed: 2]
[extra_trees: 0]
[extra_seed: 6]
[early_stopping_round: 0]
[first_metric_only: 0]
[max_delta_step: 0]
[lambda_l1: 0]
[lambda_l2: 0]
[linear_lambda: 0]
[min_gain_to_split: 0]
[drop_rate: 0.1]
[max_drop: 50]
[skip_drop: 0.5]
[xgboost_dart_mode: 0]
[uniform_drop: 0]
[drop_seed: 4]
[top_rate: 0.2]
[other_rate: 0.1]
[min_data_per_group: 100]
[max_cat_threshold: 32]
[cat_l2: 10]
[cat_smooth: 10]
[max_cat_to_onehot: 4]
[top_k: 20]
[monotone_constraints: 0,-1,-1,-1,0,0,-1,0]
[monotone_constraints_method: basic]
[monotone_penalty: 0]
[feature_contri: ]
[forcedsplits_filename: ]
[refit_decay_rate: 0.9]
[cegb_tradeoff: 1]
[cegb_penalty_split: 0]
[cegb_penalty_feature_lazy: ]
[cegb_penalty_feature_coupled: ]
[path_smooth: 0]
[interaction_constraints: ]
[verbosity: 0]
[saved_feature_importance_type: 0]
[linear_tree: 1]
[max_bin: 255]
[max_bin_by_feature: ]
[min_data_in_bin: 3]
[bin_construct_sample_cnt: 200000]
[data_random_seed: 1]
[is_enable_sparse: 0]
[enable_bundle: 1]
[use_missing: 0]
[zero_as_missing: 0]
[feature_pre_filter: 1]
[pre_partition: 0]
[two_round: 0]
[header: 0]
[label_column: ]
[weight_column: ]
[group_column: ]
[ignore_column: ]
[categorical_feature: ]
[forcedbins_filename: ]
[precise_float_parser: 1]
[parser_config_file: ]
[objective_seed: 5]
[num_class: 1]
[is_unbalance: 0]
[scale_pos_weight: 1]
[sigmoid: 0.66]
[boost_from_average: 1]
[reg_sqrt: 0]
[alpha: 0.9]
[fair_c: 1]
[poisson_max_delta_step: 0.7]
[tweedie_variance_power: 1.5]
[lambdarank_truncation_level: 30]
[lambdarank_norm: 1]
[label_gain: ]
[eval_at: ]
[multi_error_top_k: 1]
[auc_mu_weights: ]
[num_machines: 1]
[local_listen_port: 12400]
[time_out: 120]
[machine_list_filename: ]
[machines: ]
[gpu_platform_id: -1]
[gpu_device_id: -1]
[gpu_use_dp: 0]
[num_gpu: 1]

end of parameters
```


Every parameter is in the model file, even those with default values, and even those with `doc-only` in config.h, for example, `boosting` is present.

Reply:
> Just like we don't include the boosting type in the model file

Here's a minimal reproducible example showing that `boosting` does get written to the model file. Run with latest `master` (https://github.com/microsoft/LightGBM/commit/858eeb54a215ff50fd3ca0033b77d51bf48f2c1f).

```shell
# compile the CLI
mkdir build
cd ./build
cmake ..
make -j2 lightgbm

# run regression training
cd ../examples/regression
"../../lightgbm" config=train.conf

# check model file
cat LightGBM_model.txt | grep -E 'boosting'
# [boosting: gbdt]
```

And I also see lots of other parameters in there that "only affect the training", like

```text
[metric: l2]
[tree_learner: serial]
[device_type: cpu]
```

I see those same parameters when saving a model to text format with the Python package

```shell
cd ../../build
make -j2 _lightgbm
cd ..
sh ./build-python.sh install --precompile
python
```

```python
import lightgbm as lgb
from sklearn.datasets import make_regression

X, y = make_regression()
dtrain = lgb.Dataset(X, label=y)
bst = lgb.train(train_set=dtrain, params={"num_iterations": 1})
bst.save_model("out.txt")
```

.... but they aren't in `Config::SaveMembersToString()`

https://github.com/microsoft/LightGBM/blob/858eeb54a215ff50fd3ca0033b77d51bf48f2c1f/src/io/config_auto.cpp#L661

Looks like `boosting` gets special treatment in `Config::ToString()`, along with a few other parameters.

https://github.com/microsoft/LightGBM/blob/858eeb54a215ff50fd3ca0033b77d51bf48f2c1f/src/io/config.cpp#L450-L459

> *even those with `[doc-only]` in `config.h` ... are present*

Of all the parammeters `[doc-only]` or `[no-save]` in `config.h`on latest `master` (https://github.com/microsoft/LightGBM/blob/858eeb54a215ff50fd3ca0033b77d51bf48f2c1f/include/LightGBM/config.h), only those with that special handle in `Config::ToString()` are found in the model file. Not "every parameter".

```text
# explicitly added by Config::ToString()
boosting
objective
metric
tree_learner
device_type

# not in model file
config
convert_model
convert_model_language
data_sample_strategy
input_model
is_provide_training_metric
metric_freq
num_grad_quant_bins
num_iteration_predict
output_model
output_result
pred_early_stop
pred_early_stop_freq
pred_early_stop_margin
predict_contrib
predict_disable_shape_check
predict_leaf_index
predict_raw_score
quant_train_renew_leaf
save_binary
seed
snapshot_freq
start_iteration_predict
stochastic_rounding
string_config
task
use_quantized_grad
```

Reply:
@shiyu1994 @guolinke @jmoralez I think we should start saving into the model file things that:

* affect training (not logging, but the actual produced model)
* are not file paths (since those may not be portable to other environments)

Those would be useful for training continuation or for just generally inspecting models. That would mean storing the following which aren't currently stored:

```text
data_sample_strategy
num_grad_quant_bins
quant_train_renew_leaf
seed
stochastic_rounding
use_quantized_grad
```

What do you think? I'd be happy to put up a PR making those changes.

And for you @mjmckp ... could you just remove the "save that parameter value to a file" changes from this PR? So this PR's other bug fix could be merged without waiting on the outcome of this discussion.

Reply:
@jameslamb Ok, I've reverted the changes regarding persistence of `data_sample_strategy` to file, assuming that will be fixed in a separate PR.

Reply:
I agree on adding those too.

Reply:
>  I think the information is still useful for users that trained the model, as they can confirm that the model is trained from right data file.

You're right @guolinke , that reasoning makes sense to me and I agree.